scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowDoFns.scala

```diff
@@ -75,38 +80,133 @@ private[this] abstract class PredictDoFn[T, V, M <: Model[_]](
-    val model = getResource
-      .computeIfAbsent(
-        uri,
-        new Function[String, TensorFlowModel] {
-          override def apply(t: String): TensorFlowModel =
-            Models
-              .tensorFlow(uri, options)
-              .get(Duration.ofDays(Integer.MAX_VALUE))
-        }
-      )
+  override def withRunner(f: Session#Runner => V): V = f(getResource.instance().session().runner())
 
-    f(model.instance().session().runner())
+  def id(): String = {
+    val tokens = Seq("tf", uri, signatureName) ++ Seq(options.tags.asScala.mkString(";"))
+    tokens.mkString(":")
   }
 
   @Teardown
   def teardown(): Unit = {
     log.info(s"Tearing down predict DoFn $this")
-    getResource.get(uri).close()
+    getResource.close()
   }
 }
+
+object SavedBundlePredictDoFn {
+  def forRaw[T, V](
+    uri: String,
+    fetchOps: Seq[String],
+    options: TensorFlowModel.Options,
+    signatureName: String,
+    inFn: T => Map[String, Tensor[_]],
+    outFn: (T, Map[String, Tensor[_]]) => V
+  ): SavedBundlePredictDoFn[T, V] = new SavedBundlePredictDoFn[T, V](uri, signatureName, options) {
+    override def createResource(): TensorFlowModel = {
+      val model = TensorFlowLoader
+        .create(Id.create(id), uri, options, signatureName)
+        .get(Duration.ofDays(Integer.MAX_VALUE))
+      model
+    }
+
+    override def extractInput(input: T): Map[String, Tensor[_]] = inFn(input)
+
+    override def extractOutput(input: T, out: Map[String, Tensor[_]]): V = outFn(input, out)
+
+    override def outputTensorNames: Seq[String] = fetchOps
+  }
+
+  def forInput[T, V](
+    uri: String,
+    fetchOps: Option[Seq[String]],
+    options: TensorFlowModel.Options,
+    signatureName: String,
+    inFn: T => Map[String, Tensor[_]],
+    outFn: (T, Map[String, Tensor[_]]) => V
+  ): SavedBundlePredictDoFn[T, V] = new SavedBundlePredictDoFn[T, V](uri, signatureName, options) {
+    var requestedFetchOps: Map[String, String] = _
+
+    override def createResource(): TensorFlowModel = {
+      val model = TensorFlowLoader
+        .create(Id.create(id), uri, options, signatureName)
+        .get(Duration.ofDays(Integer.MAX_VALUE))
+      // by doing it here we make sure we only do it once
+      val exportedFetchOps = model.outputsNameMap().asScala.toMap
+      requestedFetchOps = fetchOps
+        .map { tensorIds =>
+          tensorIds.map { tensorId =>
+            tensorId -> exportedFetchOps(tensorId)
+          }.toMap
+        }
+        .getOrElse(exportedFetchOps)
+
+      model
+    }
+
+    override def extractInput(input: T): Map[String, Tensor[_]] = {
+      val extractedInput = inFn(input)
+      extractedInput.map {
+        case (tensorId, tensor) =>
+          getResource.inputsNameMap().get(tensorId) -> tensor
+      }.toMap
+    }
+
+    override def extractOutput(input: T, out: Map[String, Tensor[_]]): V =
+      outFn(input, requestedFetchOps.mapValues(out(_)))
```

Comment:
These days I'm up for avoid `mapValues`. can we `.iterator.map()` instead?

Reply:
Changed